File: javascript/Kamasutra.js
/*
 * Kamasutra Cipher
 *
 * @author  lellansin <lellansin@gmail.com>
 * @website http://www.lellansin.com/tutorials/ciphers
 */
(function() {

    var Kamasutra = {};

    /*
     * 生成密表
     */
    Kamasutra.initTable = function() {
        var table = [];

        // 实际应该将26个字母随机打散生成，不过这里直接用例子里面随机的
        table[0] = 'KBJHOESNWYCVI';
        table[1] = 'APMRZQGFXDULT';

        return table;
    };

    /*
     * 打印密表
     */
    Kamasutra.printTable = function(table) {
        console.log(table[0]);
        console.log(table[1]);
    };

    /*
     * 加密
     */
    Kamasutra.encrypt = function(table, words) {
        return words.replace(/([A-Z])|([a-z])/g, function(ch, upper, lower) {
            if (lower) {
                return getOpponent(table, getPosition(table, lower.toUpperCase())).toLowerCase();
            }
            return getOpponent(table, getPosition(table, upper));
        });
    };

    /*
     * 解密
     */
    Kamasutra.decrypt = function(table, text) {
        return Kamasutra.encrypt(table, text);
    };

    // 获取字母在密表中的位置
    var getPosition = function(table, ch) {
        var col = 0;
        var pos = table[col].indexOf(ch);
        if (pos >= 0)
            return [col, pos];

        pos = table[++col].indexOf(ch);
        if (pos >= 0)
            return [col, pos];
        return null;
    };

    // 根据位置获取对应另一个字母
    var getOpponent = function(table, position) {
        if (!position[0]) {
            return table[1][position[1]];
        }
        return table[0][position[1]];
    };

    /* -------------------- 测试 -------------------- */
    // 本例推算见 http://ruffnekk.stormloader.com/kamasutra_info.html

    var text = 'This is an example';

    // 生成密表 (密表就是密匙)
    var table = Kamasutra.initTable();
    Kamasutra.printTable(table);

    // 加密
    var ciphertext = Kamasutra.encrypt(table, text);
    console.log('密文:', ciphertext);

    // 解密
    console.log('解密:', Kamasutra.decrypt(table, ciphertext));

}());

File: python/Kamasutra.py
# -*- coding: utf-8 -*-
#
# Kamasutra Cipher
# 
# @author  lellansin <lellansin@gmail.com>
# @website http://www.lellansin.com/tutorials/ciphers
#

# 
# 加密
# 
def encrypt(table, words):
    cipher = ''
    for ch in words:
        if str.isalpha(ch):
            ch = getOpponent(table, ch)
        cipher += ch
    return cipher

# 
# 解密
# 
def decrypt(table, words):
    return encrypt(table, words)

def toLower(ch, flag):
    if flag:
        return ch.lower()
    else:
        return ch

#
# 获取字母在密表中的位置
#
def getPosition(table, ch):
    row = -1
    
    if ch in table[0]:
        row = 0
    elif ch in table[1]:
        row = 1

    if row != -1:
        return (row, table[row].index(ch))
    else:
        return (None, None);

# 
# 根据位置获取对应另一个字母
# 
def getOpponent(table, ch):
    flag = False
    
    if ch.islower():
        flag = True

    row, col = getPosition(table, ch.upper())

    if row == 1:
        return toLower(table[0][col], flag)
    elif row == 0:
        return toLower(table[1][col], flag)
    else:
        return ch


if __name__ == '__main__':
    # 本例推算见 http://ruffnekk.stormloader.com/kamasutra_info.html

    # 密表就是密匙
    table = [
        [ 'K', 'B', 'J', 'H', 'O', 'E', 'S', 'N', 'W', 'Y', 'C', 'V', 'I' ],
        [ 'A', 'P', 'M', 'R', 'Z', 'Q', 'G', 'F', 'X', 'D', 'U', 'L', 'T' ] ]

    # 明文
    text = "This is an example";

    # 使用密表（同时也是密匙）加密字符串
    ciphertext = encrypt(table, text)

    # 密文
    print(ciphertext)
    
    # 解密字符串
    print(decrypt(table, ciphertext))
